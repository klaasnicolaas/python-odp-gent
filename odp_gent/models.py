"""Models for Open Data Platform of Gent."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any


@dataclass
class Garage:
    """Object representing a garage."""

    garage_id: str
    name: str
    parking_type: str
    url: str

    is_open: bool
    free_parking: bool
    temporary_closed: bool

    free_space: int
    total_capacity: int
    availability_pct: float
    occupation_pct: int

    longitude: float
    latitude: float
    updated_at: datetime

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Garage:
        """Return a Garage object from a dictionary.

        Args:
            data: The data from the API.

        Returns:
            A Garage object.
        """

        attr = data["fields"]
        geo = data["geometry"]["coordinates"]
        return cls(
            garage_id=str(data.get("recordid")),
            name=attr.get("name"),
            parking_type=attr.get("type"),
            url=attr.get("urllinkaddress"),
            is_open=bool(attr.get("isopennow")),
            free_parking=bool(attr.get("freeparking")),
            temporary_closed=bool(attr.get("temporaryclosed")),
            free_space=attr.get("availablecapacity"),
            total_capacity=attr.get("totalcapacity"),
            availability_pct=round(
                (
                    float(attr.get("availablecapacity"))
                    / float(attr.get("totalcapacity"))
                )
                * 100,
                1,
            ),
            occupation_pct=attr.get("occupation"),
            longitude=geo[0],
            latitude=geo[1],
            updated_at=datetime.strptime(attr.get("lastupdate"), "%Y-%m-%dT%H:%M:%S%z"),
        )


@dataclass
class ParkAndRide:
    """Object representing a park and ride spot."""

    spot_id: str
    name: str
    parking_type: str
    url: str

    is_open: bool
    free_parking: bool
    temporary_closed: bool
    gentse_feesten: bool

    free_space: int
    total_capacity: int
    availability_pct: float
    occupation_pct: int

    longitude: float
    latitude: float
    updated_at: datetime

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> ParkAndRide:
        """Return a ParkAndRide object from a dictionary.

        Args:
            data: The data from the API.

        Returns:
            A ParkAndRide object.
        """

        def convert_bool(value: str) -> bool:
            """Convert a string to a boolean.

            Args:
                value: The string to convert.

            Returns:
                A boolean.
            """
            if value == "True":
                return True
            return False

        attr = data["fields"]
        geo = data["geometry"]["coordinates"]
        return cls(
            spot_id=str(data.get("recordid")),
            name=attr.get("name"),
            parking_type=attr.get("type"),
            url=attr.get("urllinkaddress"),
            is_open=bool(attr.get("isopennow")),
            free_parking=bool(attr.get("freeparking")),
            temporary_closed=bool(attr.get("temporaryclosed")),
            gentse_feesten=convert_bool(attr.get("gentse_feesten")),
            free_space=attr.get("availablespaces"),
            total_capacity=attr.get("numberofspaces"),
            availability_pct=round(
                (float(attr.get("availablespaces")) / float(attr.get("numberofspaces")))
                * 100,
                1,
            ),
            occupation_pct=attr.get("occupation"),
            longitude=geo[0],
            latitude=geo[1],
            updated_at=datetime.strptime(attr.get("lastupdate"), "%Y-%m-%dT%H:%M:%S%z"),
        )
