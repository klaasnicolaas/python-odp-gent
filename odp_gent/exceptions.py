"""Asynchronous Python client providing Open Data information of Gent."""


class ODPGentError(Exception):
    """Generic Open Data Platform Gent exception."""


class ODPGentConnectionError(ODPGentError):
    """Open Data Platform Gent - connection error."""
