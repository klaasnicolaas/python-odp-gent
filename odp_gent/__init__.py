"""Asynchronous Python client providing Open Data information of Gent."""

from .exceptions import ODPGentConnectionError, ODPGentError
from .models import Garage, ParkAndRide
from .odp_gent import ODPGent

__all__ = [
    "ODPGent",
    "ODPGentConnectionError",
    "ODPGentError",
    "Garage",
    "ParkAndRide",
]
