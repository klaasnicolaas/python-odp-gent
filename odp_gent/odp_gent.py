"""Asynchronous Python client providing Open Data information of Gent."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any

import aiohttp
import async_timeout
from aiohttp import hdrs
from yarl import URL

from .exceptions import ODPGentConnectionError, ODPGentError
from .models import Garage, ParkAndRide


@dataclass
class ODPGent:
    """Main class for handling data fetchting from Open Data Platform of Gent."""

    request_timeout: float = 10.0
    session: aiohttp.client.ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = hdrs.METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Gent.

        Args:
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
            A Python dictionary (text) with the response from
            the Open Data Platform API of Gent.

        Raises:
            ODPGentConnectionError: Timeout occurred while
                connecting to the Open Data Platform API.
            ODPGentError: If the data is not valid.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https", host="data.stad.gent", path="/api/records/1.0/"
        ).join(URL(uri))

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonODPGent/{version}",
        }

        if self.session is None:
            self.session = aiohttp.ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise ODPGentConnectionError(
                "Timeout occurred while connecting to the Open Data Platform API."
            ) from exception
        except (aiohttp.ClientError, socket.gaierror) as exception:
            raise ODPGentConnectionError(
                "Error occurred while communicating with Open Data Platform API."
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            raise ODPGentError(
                "Unexpected content type response from the Open Data Platform API",
                {"Content-Type": content_type, "Response": text},
            )

        return await response.json()

    async def garages(self, limit: int = 10) -> list[Garage]:
        """Get list of parking garages.

        Args:
            limit: Maximum number of garages to return.

        Returns:
            A list of Garage objects.
        """
        results: list[Garage] = []
        locations = await self._request(
            "search/",
            params={"dataset": "bezetting-parkeergarages-real-time", "rows": limit},
        )

        for item in locations["records"]:
            results.append(Garage.from_dict(item))
        return results

    async def park_and_rides(
        self, limit: int = 10, gentse_feesten: str | None = None
    ) -> list[ParkAndRide]:
        """Get list of Park and Ride locations.

        Args:
            limit: Maximum number of garages to return.
            gentse_feesten: Filter by Gentse Feesten.

        Returns:
            A list of ParkAndRide objects.
        """
        results: list[ParkAndRide] = []
        params: dict[str, Any] = {
            "dataset": "real-time-bezetting-pr-gent",
            "rows": limit,
        }

        if gentse_feesten is not None:
            params["refine.gentse_feesten"] = gentse_feesten

        locations = await self._request(
            "search/",
            params=params,
        )

        for item in locations["records"]:
            results.append(ParkAndRide.from_dict(item))
        return results

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> ODPGent:
        """Async enter.

        Returns:
            The Open Data Platform Gent object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
