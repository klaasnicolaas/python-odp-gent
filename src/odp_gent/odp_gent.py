"""Asynchronous Python client providing Open Data information of Gent."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, cast

import async_timeout
from aiohttp import ClientError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import ODPGentConnectionError, ODPGentError
from .models import BlueBike, Garage, ParkAndRide, Partago


@dataclass
class ODPGent:
    """Main class for handling data fetchting from Open Data Platform of Gent."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Gent.

        Args:
        ----
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary (text) with the response from
            the Open Data Platform API of Gent.

        Raises:
        ------
            ODPGentConnectionError: Timeout occurred while
                connecting to the Open Data Platform API.
            ODPGentError: If the data is not valid.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host="data.stad.gent",
            path="/api/records/1.0/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonODPGent/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Open Data Platform API."
            raise ODPGentConnectionError(msg) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with Open Data Platform API."
            raise ODPGentConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Open Data Platform API"
            raise ODPGentError(
                msg,
                {"Content-Type": content_type, "Response": text},
            )

        return cast(dict[str, Any], await response.json())

    async def garages(self, limit: int = 10) -> list[Garage]:
        """Get list of parking garages.

        Args:
        ----
            limit: Maximum number of garages to return.

        Returns:
        -------
            A list of Garage objects.
        """
        results: list[Garage] = []
        locations = await self._request(
            "search/",
            params={"dataset": "bezetting-parkeergarages-real-time", "rows": limit},
        )

        for item in locations["records"]:
            results.append(Garage.from_dict(item))
        return results

    async def park_and_rides(
        self,
        limit: int = 10,
        gentse_feesten: str | None = None,
    ) -> list[ParkAndRide]:
        """Get list of Park and Ride locations.

        Args:
        ----
            limit: Maximum number of garages to return.
            gentse_feesten: Filter by Gentse Feesten.

        Returns:
        -------
            A list of ParkAndRide objects.
        """
        results: list[ParkAndRide] = []
        params: dict[str, Any] = {
            "dataset": "real-time-bezetting-pr-gent",
            "rows": limit,
        }

        if gentse_feesten is not None:
            params["refine.gentse_feesten"] = gentse_feesten

        locations = await self._request(
            "search/",
            params=params,
        )

        for item in locations["records"]:
            results.append(ParkAndRide.from_dict(item))
        return results

    async def bluebikes(self) -> list[BlueBike]:
        """Get list of data from BlueBike locations.

        Returns
        -------
            A list of BlueBike objects.
        """
        results: list[BlueBike] = []

        # Data is spread over multiple datasets
        datasets: list[str] = [
            "blue-bike-deelfietsen-gent-sint-pieters-st-denijslaan",
            "blue-bike-deelfietsen-merelbeke-drongen-wondelgem",
            "blue-bike-deelfietsen-gent-sint-pieters-m-hendrikaplein",
            "blue-bike-deelfietsen-gent-dampoort",
        ]

        for dataset in datasets:
            locations = await self._request(
                "search/",
                params={"dataset": dataset},
            )

            for item in locations["records"]:
                results.append(BlueBike.from_dict(item))
        return results

    async def partago_vehicles(self, limit: int = 10) -> list[Partago]:
        """Get list of data from Partago vehicles.

        Returns
        -------
            A list of Partago objects.
        """
        results: list[Partago] = []
        vehicles = await self._request(
            "search/",
            params={"dataset": "real-time-locaties-deelwagen-partago", "rows": limit},
        )

        for item in vehicles["records"]:
            results.append(Partago.from_dict(item))
        return results

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> ODPGent:
        """Async enter.

        Returns
        -------
            The Open Data Platform Gent object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
