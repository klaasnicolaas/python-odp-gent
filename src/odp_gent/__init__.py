"""Asynchronous Python client providing Open Data information of Gent."""

from .exceptions import ODPGentConnectionError, ODPGentError
from .models import BlueBike, Garage, ParkAndRide, Partago
from .odp_gent import ODPGent

__all__ = [
    "ODPGent",
    "ODPGentConnectionError",
    "ODPGentError",
    "Garage",
    "ParkAndRide",
    "BlueBike",
    "Partago",
]
