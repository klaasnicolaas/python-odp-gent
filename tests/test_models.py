"""Test the models."""

from __future__ import annotations

from typing import TYPE_CHECKING

from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from . import load_fixtures

if TYPE_CHECKING:
    from odp_gent import BlueBike, Garage, ODPGent, ParkAndRide, Partago


async def test_all_garages(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    odp_gent_client: ODPGent,
) -> None:
    """Test all garages function."""
    aresponses.add(
        "data.stad.gent",
        "/api/records/1.0/search/",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("garages.json"),
        ),
    )
    spaces: list[Garage] = await odp_gent_client.garages()
    assert spaces == snapshot


async def test_park_and_rides(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    odp_gent_client: ODPGent,
) -> None:
    """Test park and ride spaces function."""
    aresponses.add(
        "data.stad.gent",
        "/api/records/1.0/search/",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("park_and_ride.json"),
        ),
    )
    spaces: list[ParkAndRide] = await odp_gent_client.park_and_rides()
    assert spaces == snapshot


async def test_filter_park_and_rides(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    odp_gent_client: ODPGent,
) -> None:
    """Test park and ride spaces filter function."""
    aresponses.add(
        "data.stad.gent",
        "/api/records/1.0/search/",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("park_and_ride.json"),
        ),
    )
    spaces: list[ParkAndRide] = await odp_gent_client.park_and_rides(
        gentse_feesten="True"
    )
    assert spaces == snapshot


async def test_bluebikes(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    odp_gent_client: ODPGent,
) -> None:
    """Test bluebikes function."""
    datasets: list[str] = [
        "bluebikes.json",
        "bluebikes.json",
        "bluebikes.json",
        "bluebikes.json",
    ]
    for dataset in datasets:
        aresponses.add(
            "data.stad.gent",
            "/api/records/1.0/search/",
            "GET",
            aresponses.Response(
                status=200,
                headers={"Content-Type": "application/json"},
                text=load_fixtures(dataset),
            ),
        )
    bluebikes: list[BlueBike] = await odp_gent_client.bluebikes()
    assert bluebikes == snapshot


async def test_partago(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    odp_gent_client: ODPGent,
) -> None:
    """Test partago function."""
    aresponses.add(
        "data.stad.gent",
        "/api/records/1.0/search/",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("partago.json"),
        ),
    )
    partago_vehicles: list[Partago] = await odp_gent_client.partago_vehicles()
    assert partago_vehicles == snapshot
