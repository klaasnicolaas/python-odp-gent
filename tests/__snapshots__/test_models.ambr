# serializer version: 1
# name: test_all_garages
  list([
    Garage(garage_id='72dc32bd8fe81dae0b9602a038aaeea392d64c30', name='Sint-Michiels', parking_type='carPark', url='https://stad.gent/nl/mobiliteit-openbare-werken/parkeren/parkings-gent/parking-sint-michiels', is_open=True, free_parking=False, temporary_closed=False, free_space=181, total_capacity=315, availability_pct=57.5, occupation_pct=42, longitude=3.719727740000235, latitude=51.05311004768356, updated_at=datetime.datetime(2022, 7, 15, 13, 36, 45, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
    Garage(garage_id='713cf9789dc799f115348ac45bbae7a4df89a13b', name='Reep', parking_type='carPark', url='https://stad.gent/nl/mobiliteit-openbare-werken/parkeren/parkings-gent/parking-reep', is_open=True, free_parking=False, temporary_closed=False, free_space=174, total_capacity=439, availability_pct=39.6, occupation_pct=60, longitude=3.7298928896398924, latitude=51.05215838523151, updated_at=datetime.datetime(2022, 7, 15, 13, 36, 45, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
  ])
# ---
# name: test_bluebikes
  list([
    BlueBike(spot_id=72, name='Station Gent-Dampoort', spot_type=1, bikes_in_use=15, bikes_available=45, last_update=datetime.datetime(2023, 5, 6, 20, 34, 59, tzinfo=<DstTzInfo 'Europe/Brussels' LMT+0:18:00 STD>), longitude=3.740194, latitude=51.056083),
    BlueBike(spot_id=72, name='Station Gent-Dampoort', spot_type=1, bikes_in_use=15, bikes_available=45, last_update=datetime.datetime(2023, 5, 6, 20, 34, 59, tzinfo=<DstTzInfo 'Europe/Brussels' LMT+0:18:00 STD>), longitude=3.740194, latitude=51.056083),
    BlueBike(spot_id=72, name='Station Gent-Dampoort', spot_type=1, bikes_in_use=15, bikes_available=45, last_update=datetime.datetime(2023, 5, 6, 20, 34, 59, tzinfo=<DstTzInfo 'Europe/Brussels' LMT+0:18:00 STD>), longitude=3.740194, latitude=51.056083),
    BlueBike(spot_id=72, name='Station Gent-Dampoort', spot_type=1, bikes_in_use=15, bikes_available=45, last_update=datetime.datetime(2023, 5, 6, 20, 34, 59, tzinfo=<DstTzInfo 'Europe/Brussels' LMT+0:18:00 STD>), longitude=3.740194, latitude=51.056083),
  ])
# ---
# name: test_filter_park_and_rides
  list([
    ParkAndRide(spot_id='3f38ec867beb355b795247f416c782cdb5695338', name='P+R Wondelgem', parking_type='offStreetParkingGround', url='https://stad.gent/nl/mobiliteit-openbare-werken/parkeren/park-and-ride-pr/pr-wondelgem-industrieweg', is_open=True, free_parking=True, temporary_closed=False, gentse_feesten=False, free_space=18, total_capacity=62, availability_pct=29.0, occupation_pct=70, longitude=3.71138627257, latitude=51.0985019636, updated_at=datetime.datetime(2022, 7, 15, 13, 30, 42, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
    ParkAndRide(spot_id='ab99b16b828b42be404ad006b86ffd476ef2a708', name='P+R Bourgoyen', parking_type='offStreetParkingGround', url='https://stad.gent/nl/mobiliteit-openbare-werken/parkeren/park-and-ride-pr/pr-bourgoyen', is_open=True, free_parking=True, temporary_closed=False, gentse_feesten=True, free_space=211, total_capacity=250, availability_pct=84.4, occupation_pct=15, longitude=3.68175794804, latitude=51.0674048936, updated_at=datetime.datetime(2022, 7, 15, 14, 31, 1, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
  ])
# ---
# name: test_park_and_rides
  list([
    ParkAndRide(spot_id='3f38ec867beb355b795247f416c782cdb5695338', name='P+R Wondelgem', parking_type='offStreetParkingGround', url='https://stad.gent/nl/mobiliteit-openbare-werken/parkeren/park-and-ride-pr/pr-wondelgem-industrieweg', is_open=True, free_parking=True, temporary_closed=False, gentse_feesten=False, free_space=18, total_capacity=62, availability_pct=29.0, occupation_pct=70, longitude=3.71138627257, latitude=51.0985019636, updated_at=datetime.datetime(2022, 7, 15, 13, 30, 42, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
    ParkAndRide(spot_id='ab99b16b828b42be404ad006b86ffd476ef2a708', name='P+R Bourgoyen', parking_type='offStreetParkingGround', url='https://stad.gent/nl/mobiliteit-openbare-werken/parkeren/park-and-ride-pr/pr-bourgoyen', is_open=True, free_parking=True, temporary_closed=False, gentse_feesten=True, free_space=211, total_capacity=250, availability_pct=84.4, occupation_pct=15, longitude=3.68175794804, latitude=51.0674048936, updated_at=datetime.datetime(2022, 7, 15, 14, 31, 1, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
  ])
# ---
# name: test_partago
  list([
    Partago(name='Partago Quark', vehicle_type=Vehicle(brand='Nissan', mode='e-NV200 2018', fuel_type='electric', transmission_type='automatic'), picture_url='https://firebasestorage.googleapis.com/v0/b/sizzling-torch-8026.appspot.com/o/images%2Fplaces%2Fquark.png?alt=media&token=281f4992-fb33-49e3-adbf-1cd51f5a45d5', station_type='free floating', last_update=datetime.datetime(2023, 5, 7, 12, 0, 8, 601000, tzinfo=<UTC>), latitude=51.047994818648135, longitude=3.717662708201117),
    Partago(name='Partago Theo', vehicle_type=Vehicle(brand='Renault', mode='ZOE 50', fuel_type='electric', transmission_type='automatic'), picture_url='https://firebasestorage.googleapis.com/v0/b/sizzling-torch-8026.appspot.com/o/images%2Fplaces%2Fzoe_66percent.jpg?alt=media&token=948566aa-abf0-4909-8e4d-62177d686877', station_type='free floating', last_update=datetime.datetime(2023, 5, 7, 12, 0, 8, 601000, tzinfo=<UTC>), latitude=51.044021518441454, longitude=3.7530556155420487),
    Partago(name='Partago Rox', vehicle_type=Vehicle(brand='Renault', mode='Zoé', fuel_type='electric', transmission_type='automatic'), picture_url='https://firebasestorage.googleapis.com/v0/b/sizzling-torch-8026.appspot.com/o/images%2Fplaces%2Fzoe_66percent.jpg?alt=media&token=948566aa-abf0-4909-8e4d-62177d686877', station_type='free floating', last_update=datetime.datetime(2023, 5, 7, 12, 0, 8, 601000, tzinfo=<UTC>), latitude=51.03416087693723, longitude=3.7441249969609345),
    Partago(name='Partago Maxwell', vehicle_type=Vehicle(brand='Renault', mode='Zoé', fuel_type='electric', transmission_type='automatic'), picture_url='https://firebasestorage.googleapis.com/v0/b/sizzling-torch-8026.appspot.com/o/images%2Fplaces%2Fzoe_66percent.jpg?alt=media&token=948566aa-abf0-4909-8e4d-62177d686877', station_type='free floating', last_update=datetime.datetime(2023, 5, 7, 12, 0, 8, 601000, tzinfo=<UTC>), latitude=51.0628833078908, longitude=3.7313283733281333),
    Partago(name='Partago Wez', vehicle_type=Vehicle(brand='Renault', mode='Zoé', fuel_type='electric', transmission_type='automatic'), picture_url='https://firebasestorage.googleapis.com/v0/b/sizzling-torch-8026.appspot.com/o/images%2Fplaces%2Fzoe_66percent.jpg?alt=media&token=948566aa-abf0-4909-8e4d-62177d686877', station_type='free floating', last_update=datetime.datetime(2023, 5, 7, 12, 0, 8, 601000, tzinfo=<UTC>), latitude=51.0443421974434, longitude=3.760778443329769),
  ])
# ---
